src/poke_env/player/openai_api.py

```diff
@@ -0,0 +1,537 @@
+        if action == -1:
+            return ForfeitBattleOrder()
+        return self.user_funcs.action_to_move(action, battle)
+
+    def _battle_finished_callback(
+        self, battle: AbstractBattle
+    ) -> None:  # pragma: no cover
+        to_put = self.user_funcs.embed_battle(battle)
+        asyncio.run_coroutine_threadsafe(
+            self.observations.async_put(to_put), asyncio.get_event_loop()
+        )
+
+
+class OpenAIGymEnv(Env, ABC):  # pyre-ignore
+
+    _INIT_RETRIES = 100
+    _TIME_BETWEEN_RETRIES = 0.5
+    _SWITCH_CHALLENGE_TASK_RETRIES = 30
+    _TIME_BETWEEN_SWITCH_RETIRES = 1
+
+    def __init__(
+        self,
+        player_configuration: Optional[PlayerConfiguration] = None,
+        *,
+        avatar: Optional[int] = None,
+        battle_format: str = "gen8randombattle",
+        log_level: Optional[int] = None,
+        save_replays: Union[bool, str] = False,
+        server_configuration: Optional[
+            ServerConfiguration
+        ] = LocalhostServerConfiguration,
+        start_timer_on_battle_start: bool = False,
+        start_listening: bool = True,
+        team: Optional[Union[str, Teambuilder]] = None,
+        start_challenging: bool = False,
+    ):
+        if not asyncio.get_event_loop() == THREAD_LOOP:
+            raise RuntimeError(
+                "You are attempting to create an OpenAI agent without "
+                "the required context. "
+                "Remember to create all the players you will use "
+                "with the OpenAI API inside the context "
+                "EnvLoop. If unsure, use the following:\n"
+                "if __name__ == '__main__':\n"
+                "\twith EnvLoop():\n"
+                "\t\tmain()"
+            )
+        self.agent = _AsyncPlayer(
+            self,
+            username=self.__class__.__name__,
+            player_configuration=player_configuration,
+            avatar=avatar,
+            battle_format=battle_format,
+            log_level=log_level,
+            max_concurrent_battles=1,
+            save_replays=save_replays,
+            server_configuration=server_configuration,
+            start_timer_on_battle_start=start_timer_on_battle_start,
+            start_listening=start_listening,
+            team=team,
+        )
+        self.battle_format = battle_format
+        self.actions = self.agent.actions
+        self.observations = self.agent.observations
+        self.action_space = Discrete(self.action_space_size())
+        self.observation_space = self.describe_embedding()
+        self.current_battle: Optional[AbstractBattle] = None
+        self.last_battle: Optional[AbstractBattle] = None
+        self._keep_challenging: bool = False
+        self.challenge_task = None
+        if start_challenging:
+            self._keep_challenging = True
+            self.challenge_task = asyncio.run_coroutine_threadsafe(
+                self.challenge_loop(), asyncio.get_event_loop()
```

Comment:
Would it be possible to:
- test what the current event loop is, and if necessary change it to `THREAD_LOOP` and
- then pass `THREAD_LOOP` directly, and therefore avoid the context manager?

Reply:
The problem with this is that it wouldn't work: for this to work every single agent that will interact with the `OpenAIGymEnv` HAS to be listening on the `THREAD_LOOP`. This can be achieved by what you proposed plus writing in the docs that every agent that needs to interact with the `OpenAIGymEnv`  needs to be creted with `start_listening=False` and then to have it listen on the correct loop. I think the context manager simplifies this and allows to be sure that when outside of the context the main event loop (if present) is used

Reply:
That makes sense. Let me think about it a bit more - I'll get back to you tomorrow

Reply:
Hey,
I get where you're going with this and I think that while it's possible to implement it wouldn't remove the context manager: when you need to call an async method of the player you would need to call it using:
```
task = asyncio.run_coroutine_threadsafe(player.battle_against(opponent, 1), player.loop)
```
Same thing for the unified thread:
```
task = asyncio.run_coroutine_threadsafe(player.battle_against(opponent, 1), BACKGROUND_LOOP)
```
And then collect the result using:
```
task.result()
```
While this is true also for the context manager:
```
task = asyncio.run_coroutine_threadsafe(player.battle_against(opponent, 1), asyncio.get_event_loop())
```
the context manager would allow users to call other players normally with `await` when outside of it, so it would retain backward compatibility.

Reply:
I think the best solution is the shared background thread then. I'll start working on it tomorrow

Reply:
Hey @hsahovic,
I opened a new draft with the changes you requested but could you consider merging #264?
I explained the rationale for it in the PR

Thank you

Reply:
I merged #264 but I worry medium+ might not actually be enough - let me know if you think we should go another step up in RAM

Reply:
Hey @hsahovic,
take a look at #263.
It should do what you wanted without actually changing the API :)

Let me know if the API are ok so I can get on writing the tutorials and the documentation